lib/pharos/up_command.rb

```diff
@@ -77,27 +77,39 @@ def configure(config)
       master_hosts = master_hosts(config)
       signal_usage_error 'No master hosts defined' if master_hosts.empty?
       signal_usage_error 'Only one host can be in master role' if master_hosts.size > 1
-
-      begin
-        start_time = Time.now
-        puts pastel.green("==> Sharpening tools ...")
-        load_phases
-        puts pastel.green("==> Starting to craft cluster ...")
-        validate_hosts(config)
-        # set workdir to the same dir where config was loaded from
-        # so that the certs etc. can be referenced more easily
-        Dir.chdir(config_yaml.dirname) do
-          handle_masters(master_hosts[0], config)
-          handle_workers(master_hosts[0], worker_hosts(config), config)
-          handle_addons(master_hosts[0], config.addons)
-        end
-        craft_time = Time.now - start_time
-        puts pastel.green("==> Cluster has been crafted! (took #{humanize_duration(craft_time.to_i)})")
-        puts "    You can connect to the cluster with kubectl using:"
-        puts "    export KUBECONFIG=~/.pharos/#{master_hosts[0].address}"
-      ensure
-        Pharos::SSH::Client.disconnect_all
+      master_host = master_hosts[0]
+
+      start_time = Time.now
+      puts pastel.green("==> Sharpening tools ...")
+      load_phases
+      puts pastel.green("==> Starting to craft cluster ...")
+
+      # set workdir to the same dir where config was loaded from
+      # so that the certs etc. can be referenced more easily
+      Dir.chdir(config_yaml.dirname) do
+        handle_phase(Phases::ValidateHost, config.hosts, config: config)
+        handle_phase(Phases::ConfigureHost, config.hosts, config: config)
+
+        handle_phase(Phases::ConfigureMaster, master_hosts, config: config, master: master_host)
+        handle_phase(Phases::ConfigureClient, master_hosts, config: config, master: master_host)
+        handle_phase(Phases::ConfigureDNS, master_hosts, config: config, master: master_host)
+        handle_phase(Phases::ConfigureNetwork, master_hosts, config: config, master: master_host)
+        handle_phase(Phases::LabelNode, master_hosts, config: config, master: master_host)
+        handle_phase(Phases::StoreClusterYAML, master_hosts, master: master_host, config_content: config_yaml.read(ENV.to_h))
+
+        handle_phase(Phases::ConfigureKubelet, worker_hosts(config), config: config, master: master_host)
```

Comment:
The `ConfigureKubelet` phase could run in parallel across both the master + worker hosts, *if* this is an initial install... the upgrade case is a little more complicated.

That logic is currently handled within the `ConfigureMaster` phase, it might make sense to deal with here instead?